cpp/dolfin/function/Expression.h

```diff
@@ -94,35 +103,50 @@ class Expression
                        Eigen::RowMajor>
   compute_point_values(const mesh::Mesh& mesh) const;
 
-  /// Evaluate method
+  /// Evaluate at given point in given cell
   ///
-  /// Signature of the method accepts:
-  /// @param values
-  ///        Pointer to row major 2D C-style array of `PetscScalar`.
-  ///        The array has shape=(number of points, value size) and has to
-  ///        be filled with custom values in the function body.
-  /// @param x
-  ///        Pointer to a row major C-style 2D array of `double`.
-  ///        The array has shape=(number of points, geometrical dimension)
-  ///        and represents array of points in physical space at which the
-  ///        Expression is being evaluated.
-  /// @param cell_idx
-  ///        Pointer to a 1D C-style array of `int`. It is an array
-  ///        of indices of cells where points are evaluated. Value -1 represents
-  ///        cell-independent eval function
-  /// @param num_points
-  ///        Number of points where expression is evaluated
-  /// @param value_size
-  ///        Size of expression value
-  /// @param gdim
-  ///        Geometrical dimension of physical point where expression
-  ///        is evaluated
-  /// @param num_cells
-  ///        Number of cells
+  /// @param    values (Eigen::Ref<Eigen::VectorXd>)
+  ///         The values at the point.
+  /// @param    x (Eigen::Ref<const Eigen::VectorXd>)
+  ///         The coordinates of the point.
+  /// @param    cell (mesh::Cell)
+  ///         The cell which contains the given point.
+  virtual void eval(Eigen::Ref<Eigen::Array<PetscScalar, Eigen::Dynamic,
+                                            Eigen::Dynamic, Eigen::RowMajor>>
+                        values,
+                    const Eigen::Ref<const EigenRowArrayXXd> x,
+                    const dolfin::mesh::Cell& cell) const;
```

Comment:
Here should be vector of `Cell` to keep all arguments vectorise-able. Or vector of cell indices as in C signature.

Reply:
That will come later (and will be an index rather than cell).

Reply:
Actually, the whole cell business is broken in the context of vectorisation because we would need to make `x` a three-dimensional array to know which points belong to which cell.

Reply:
Could we for non-mixed meshes just put them together and say `x[0:3,:]` is for first cell, `x[3:6, :]` for second etc. First cell would have index `cell_idx[0]`, second `cell_idx[1]`, etc

Reply:
If we keep the cell id, it should have the same number of rows as x, and for each x give the cell. This is general and clear - no assumption in fixed number of points per cell.